File: backend/models/data_source.py
import enum

from .base import Base

from sqlalchemy import ForeignKey, JSON, String, Integer, Enum, Boolean
from sqlalchemy.orm import Mapped, mapped_column


class DataSourceType(enum.Enum):
    postgresql = "POSTGRESQL"
    mysql = "MYSQL"


class DataSource(Base):
    name: Mapped[str] = mapped_column(String, nullable=False)
    created_by_id: Mapped[int] = mapped_column(ForeignKey("user.id"))
    credentials: Mapped[dict] = mapped_column(JSON, nullable=False)
    type: Mapped[DataSourceType] = mapped_column(Enum(DataSourceType))
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)

File: frontend/types/data_source.ts
import { BaseSchema } from '@/types/base'

export interface CredentialsSchema {
  host: string
  username: string
  password: string
  db: string
  port: string
}

export enum DataSourceType {
  postgresql = 'POSTGRESQL',
  mysql = 'MYSQL',
}

export interface CheckConnectionSchema {
  credentials: CredentialsSchema
  type: DataSourceType | string
}

export interface AddDataSourceSchema extends CheckConnectionSchema {
  name: string
}

export interface CheckConnectionResult {
  connection_result: boolean
  connection_error?: string
}

export interface DataSource extends BaseSchema {
  credentials: CredentialsSchema,
  id: number,
  created_by_id: number,
  name: string,
  type: DataSourceType
}
